src/kOS.Safe/Encapsulation/ScalarValue.cs

```diff
@@ -0,0 +1,456 @@
+                {
+                    double d = GetDoubleValue();
+                    if (double.IsInfinity(d) || double.IsNaN(d)) return false;
+                    return true;
+                }
+                return false;
+            }
+        }
+
+        public object Value { get { return internalValue; } }
+
+        protected ScalarValue()
+        {
+            InitializeSuffixes();
+        }
+
+        public void InitializeSuffixes()
+        {
+            AddSuffix("ISINT", new Suffix<bool>(() => IsInt));
+            AddSuffix("ISDOUBLE", new Suffix<bool>(() => IsDouble));
+            AddSuffix("ISVALID", new Suffix<bool>(() => IsValid));
+        }
+
+        public static ScalarValue Create(object value)
+        {
+            if (value is float)
+                value = Convert.ToDouble(value);
+            if (value is double)
+            {
+                bool inBounds = Int32.MinValue < (double)value && (double)value < Int32.MaxValue;
+                if (inBounds && !double.IsNaN((double)value))
+                {
+                    // Convert the double to an int, and check and see if they are still equal.
+                    // if so, treat the double as if it was an int.
+                    int intPart = Convert.ToInt32(value);
+                    if ((double)value == intPart)
+                    {
+                        return new ScalarIntValue(intPart);
+                    }
+                }
+                return new ScalarDoubleValue((double)value);
+            }
+            else if (value is int)
+            {
+                return new ScalarIntValue((int)value);
+            }
+            else if (value is ScalarValue)
+            {
+                return ScalarValue.Create(((ScalarValue)value).Value);
+            }
+            else
+            {
+                throw new kOS.Safe.Exceptions.KOSException(string.Format("Failed to set scalar value.  Passed type {0}, expected Double or Int", value.GetType().Name));
+            }
+        }
+
+        public int GetIntValue()
+        {
+            return Convert.ToInt32(Value);
+        }
+
+        public double GetDoubleValue()
+        {
+            return Convert.ToDouble(Value);
+        }
+
+        public override string ToString()
+        {
+            if (IsInt) return GetIntValue().ToString();
+            else if (IsDouble) return GetDoubleValue().ToString();
+            return "NaN";
+        }
+
+        public override bool Equals(object obj)
+        {
+            if (obj == null) return false;
+            if (obj is ScalarValue)
+            {
+                ScalarValue val = obj as ScalarValue;
+                if (val != null)
+                {
+                    if (this.IsInt && val.IsInt)
+                    {
+                        return this.GetIntValue() == val.GetIntValue();
+                    }
+                    return this.GetDoubleValue() == val.GetDoubleValue();
```

Comment:
It won't be a problem.

32-bit floats can store an accurate integer up to (2^24 - 1).
64-bit floats can store an accurate integer up to (2^53 - 1)

The standard floating point number formats are essentially just scientific notation, with the following two important differences:

1 - The mantissa is expressed only as a fractional part between 0 and 1, rather than with having one digit to the left of the point as it usually is with scientific notation. (i.e. normal scientific notation would express the number 123 as 1.23e2, while the way computer numbers do it would express it more like 123e3, with the nomenclature using a shorthand that there's an implied "0." preceding the integer 123, that just isn't shown in the format.  This nomenclature allows the format to never mention a decimal place, and just write it out with integers.)
2 - Except it's not in base 10 of course, it's base 2.  So the nomenclature 123e3 mentioned above would actually be more like this: 1111011e111 ( 0.123, (stored as 1111011), shifted 7 (stored as 111) binary places).
3- It's a bit more complex when the number is negative because it does negatives using 2's compliment representation of the mantissa integer, but that's the general idea.

Thus the largest exact integer it can hold is just however many binary digits are dedicated to the mantissa part.  Looking up the IEEE standards on it, the 64 bits of a double are allocated as 10 bits for the exponent part, and 54 bits for the mantissa integer part (one of which is the sign bit so it's really 53 bits for the max int, minus 1 for how 2s complement works, thus the value (2^53 - 1) I mentioned above.)

I went through working this out a while back when I was worried about how we were letting people index arrays and lists using doubles instead of ints, and how I thought it would break because it can't actually convert down into the integers the array indeces expect, but then I did that above work and realized the integers would have to get enormous before it would truly break.